### ISSUE TITLE
Build Error due to Mismatch Between `TargetFrameworkVersion` and `targetSdkVersion`

### ISSUE THREAD

**Issue by npostma** [dotnet/maui#21820]
### Description

I have encountered an issue where the `TargetFrameworkVersion` (Android API level 34) is reported to be higher than the targetSdkVersion (28), leading to a build failure with the following error message:

The TargetFrameworkVersion (Android API level 34) is higher than the targetSdkVersion (28). Please increase the `android:targetSdkVersion` in the `AndroidManifest.xml` so that the API levels match.

This error occurs despite my project being configured to target API 33. The relevant configuration in my .csproj file is as follows:
<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">33.0</SupportedOSPlatformVersion>

For the targetframework, is use:
<TargetFrameworks>net8.0-ios;net8.0-android</TargetFrameworks>

Furthermore, attempting to manually add a uses-sdk entry in the AndroidManifest.xml results in another error indicating that two uses-sdk entries are present, suggesting that one is being auto-generated during the build process.

Additionally, due to this configuration issue, I am unable to publish my app, as it fails to meet Google's requirement to target at least Android API 33.

Adding <TargetFrameworkVersion>v12.0</TargetFrameworkVersion>  to the .csproj does change the error to:

The TargetFrameworkVersion (Android API level 34) is higher than the targetSdkVersion (28). Please increase the `android:targetSdkVersion` in the `AndroidManifest.xml` so that the API levels match.
The minSdkVersion (33) is greater than the targetSdkVersion. Please change the value such that the minSdkVersion is less than or equal to the targetSdkVersion (28).

So that seems to make it more messy.

**Steps to Reproduce:**

Set the SupportedOSPlatformVersion in the .csproj file.
Attempt to build the project.
Observe the error regarding mismatched API levels.
Attempt to add uses-sdk in the AndroidManifest.xml.

**Expected Behavior:**
The project should correctly interpret the API levels specified in the .csproj file without generating conflicting or incorrect uses-sdk entries in the AndroidManifest.xml.

**Actual Behavior:**
The build process incorrectly references API levels, leading to a mismatch error and preventing app publication due to outdated target SDK specifications.

Please advise on a resolution or provide guidance on how to correctly configure the project to align the TargetFrameworkVersion with the targetSdkVersion as per the new specifications.

### Version with bug

8.0.20 SR4

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.14 SR3.1

### Affected platforms

Android

### Did you find any workaround?

No, i tried VS 2022 and the VS 2022 preview. Both behave the same.

**Comment by Redth** [comment]
There's generally 3 "sdk" versions to consider for Android development (two of which can appear in your AndroidManifest.xml file):

## 1. minimumSdkVersion

This can appear in the `<uses-sdk` element of your AndroidManifest.xml file.  It defines the oldest device your library/app supports running on.  Generally you want to keep this lower, and we currently in .NET 8 support as far back as API 21 (Android 5.0 / Lollipop).  In your .NET Android / MAUI project you can specify the value by using the `<SupportedOSPlatformVersion>21.0</SupportedOSPlatformVersion>` msbuild property.

## 2. compileSdkVersion

This one is not part of the manifest file, but this concept in a native / gradle android project maps to .NET's `TargetPlatformVersion`, which is the API level you are compiling against.  Now it's important to note that the `TargetFramework` version has an implicit default value for the `TargetPlatformVersion`.   I believe the default for `net8.0-android` is `34.0`.  So when you specify `net8.0-android` it implies `net8.0-android34.0` as the full target framework which means you are compiling against API level 34.

This really controls the API surface available to your project to write code and compile against.  However, even though you can see, use, and compile against all of the API's in API Level 34, you still need to guard your 'runtime' usage of any API's that are not available on API levels older than your `TargetPlatformVersion` all the way back to your `SupportedOSPlatformVersion` if the two are not the same in your project.

If .NET for Android supports API level 35 before .NET 9, you should be able to explicitly use it by specifying `net8.0-android35.0`.  You can't really go back to _older_ API levels as the target platform version though, and each major .NET version may change what the implicit version is.

`TargetFrameworkVersion` is `v8.0` for .NET 8 in this case, and bumping that to `v12.0` is invalid since this version corresponds to the .NET SDK version not the Android API level.

## 3. targetSdkVersion

General advice here is unless you have good reason and know _why_ you would want to do so, you should leave this one alone and allow it to implicitly match your `compileSdkVersion` / `TargetPlatformVersion`.  There are some scenarios that Google advices to use this value with a different version than the SDK you are compiling against (mostly it's to opt into older platform behaviours while still compiling against the newer SDK).

There is no mapping to .NET that I'm aware of (@jonathanpeppers can correct me if I'm wrong) and if you do want to set this value you can manually set it in your AndroidManifest.xml file in your project.


------


Now, the question is, _why_ are you trying to change the targetSdkVersion for your project? If I had to guess, it might be to try and make your life easier in terms of knowing what API's are available to you at compile time so that you don't use API's newer than a particular API level?  If that's the case, the best way to achieve this would be to look at your code and resolve any `CA1416` warnings (perhaps making them warn as error).  You'll see something like: `This call site is reachable on: 'Android' 21.0 and later.  'Xyz' is only supported on: 'android' 34.0 and later.`.

Hope that helps resolve some confusion around these versions and how to use them!

**Comment by jonathanpeppers** [comment]
Generally, you shouldn't use the `AndroidManifest.xml` anymore, and use the new MSBuild properties instead:

* https://learn.microsoft.com/en-us/dotnet/maui/migration/android-projects?view=net-maui-8.0#changes-to-androidmanifestxml

I would just put what the project template has:

```xml
<Project>
  <PropertyGroup>
    <!-- 34 is implicit (net8.0-android34), leave this blank so you get 35 when you move to .NET 9 -->
    <TargetFramework>net8.0-android</TargetFramework>
    <!-- Only reason to change this is if you want a *higher* minimum and maybe you get some C# warnings about the APIs you are using -->
    <SupportedOSPlatformVersion>21</SupportedOSPlatformVersion>
  </PropertyGroup>
</Project>
```

**Comment by npostma** [comment]
I am not sure why this is closed. Because what you tolt me, i knew, and this is exactly what is not working. If i use the defaults:
```xml
<PropertyGroup>
	<TargetFrameworks>net8.0-ios;net8.0-android</TargetFrameworks>
	<OutputType>Exe</OutputType>
	<RootNamespace>MyNamespace</RootNamespace>
	<UseMaui>true</UseMaui>
	<SingleProject>true</SingleProject>
	<ImplicitUsings>enable</ImplicitUsings>

	<!-- Display name -->
	<ApplicationTitle>MyApp</ApplicationTitle>

	<!-- App Identifier -->
	<ApplicationId>com.companyname.myapp</ApplicationId>
	<ApplicationIdGuid>some_uuid</ApplicationIdGuid>

	<!-- Versions -->
	<ApplicationDisplayVersion>1.0</ApplicationDisplayVersion>
	<ApplicationVersion>1</ApplicationVersion>

	<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'">14.4</SupportedOSPlatformVersion>
	<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">33.0</SupportedOSPlatformVersion>
	<BuildWithMSBuildOnMono>false</BuildWithMSBuildOnMono>
	<Configurations>Release;Debug</Configurations>
	<PlatformTarget>AnyCPU</PlatformTarget>
</PropertyGroup>
```

I get the error `The TargetFrameworkVersion (Android API level 34) is higher than the targetSdkVersion (28). Please increase the android:targetSdkVersion in the AndroidManifest.xml so that the API levels match.` 

Also the produced archive cant be uploaded to google, because they want API level 33 as a minimum and it says that it targets API28 which i never target anywhere ..... so what am i missing?

Error google console:

```Your app currently targets API level 28 and must target at least API level 33 to ensure the app is based on the latest APIs optimized for security and performance. More information```

Setting those other options where an attempt to fix the issue.

**Comment by jonathanpeppers** [comment]
What are the contents of your `AndroidManifest.xml`? is there a `<uses-sdk/>` element you should remove?

**Comment by andreas-spindler-mw** [comment]
Having the same problem. Build is rejected by PlayStore with the exact same error message. TargetSdk should be 33, but is actually 31. See below.

`dotnet --version -> 8.0.204`

My csproj file contains (among other stuff, but this is all about versioning):
```
<TargetFrameworks>net8.0-android;net8.0-ios</TargetFrameworks>
<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">26.0</SupportedOSPlatformVersion>
```

The generated AndroidManifest after a `clean/rebuild` in `obj/Debug/net8.0-android/` contains:
```
<uses-sdk android:minSdkVersion="26" android:targetSdkVersion="31" />
```
If I add the following line to the csproj:
```
<TargetPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">34</TargetPlatformVersion>
```
Then I get the above mentioned build warning and the targetSdkVersion is still 31.
Any advice?

**Comment by jonathanpeppers** [comment]
So a NuGet package (with an .aar) must have an `AndroidManifest.xml` inside.

If you force your `AndroidManifest.xml` to:
```
<uses-sdk android:minSdkVersion="21" android:targetSdkVersion="34" />
```
Does it solve the problem? A `.binlog` might show what package it is: https://aka.ms/binlog

**Comment by npostma** [comment]
@jonathanpeppers Hi, thanks for responding. 

`What are the contents of your AndroidManifest.xml? is there a <uses-sdk/> element you should remove?`
Like i said in the original post, I do not have a <uses-sdk/> by default. attempting to manually add a uses-sdk entry in the AndroidManifest.xml results in another error indicating that two uses-sdk entries are present, suggesting that one is being auto-generated during the build process.

So i use the XML sample above and do not have a <uses-sdk/>. Then if i clean the solution, delete bin and obj folders i get the error mentioned above. `The TargetFrameworkVersion (Android API level 34) is higher than the targetSdkVersion (28). Please increase the android:targetSdkVersion in the AndroidManifest.xml so that the API levels match.` and the generated .abb file in the google console gives me the error `Your app currently targets API level 28 and must target at least API level 33 to ensure the app is based on the latest APIs optimized for security and performance.`

If i check the generated AndroidManifes.xml (the one that is created in the build process) it has this entry:

```
<!--suppress UsesMinSdkAttributes-->
<uses-sdk android:minSdkVersion="33" />
```

**Comment by andreas-spindler-mw** [comment]
Experiencing exactly the same what @npostma  describes.

**Comment by npostma** [comment]
Some additional information: If i search for `targetSdkVersion` in the binlog i find this section:

```text
Output Property: _AndroidPackage=com.Company.MyApp.Experimental
Done executing task "GetAndroidPackageName".
Done building target "_GetAndroidPackageName" in project "MyApp.csproj".
Target "_GetJavaPlatformJar" in file "C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.95\tools\Xamarin.Android.Tooling.targets" from project "C:\src\sc_mc\MyApp\MyApp\MyApp.csproj" (target "_ValidateAndroidPackageProperties" depends on it):
Using "GetJavaPlatformJar" task from assembly "C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.95\tools\Xamarin.Android.Build.Tasks.dll".
Task "GetJavaPlatformJar"
Task Parameter:DesignTimeBuild=False
Task Parameter:BuildingInsideVisualStudio=True
Task Parameter:TargetFramework=net8.0-android
Task Parameter:AndroidSdkDirectory=C:\Program Files (x86)\Android\android-sdk
Task Parameter:SupportedOSPlatformVersion=33.0
Task Parameter:AndroidManifest=C:\src\sc_mc\MyApp\MyApp\Platforms\Android\AndroidManifest.xml
Task Parameter:AndroidSdkPlatform=34
Output Property: JavaPlatformJarPath=C:\Program Files %28x86%29\Android\android-sdk\platforms\android-34\android.jar
Output Property: _AndroidTargetSdkVersion=34
Done executing task "GetJavaPlatformJar".
Done building target "_GetJavaPlatformJar" in project "MyApp.csproj".
```

Here you can see targetSdkVersion is set to 33. But still at the end of the process i get the error and cant use the archive in the google console for publishing.

**Comment by jonathanpeppers** [comment]
Did anyone share a `.binlog`? https://aka.ms/binlog

The problem isn't in this repo, so you could file at: https://github.com/xamarin/xamarin-android/issues

**Comment by Equabyte** [comment]
Guys not sure who closed this btw issue seems coming from the AndroidManifest.xml file automatically generated and stored inside net8.0-android > android which contains something like this 
  <uses-sdk android:minSdkVersion="23" android:targetSdkVersion="28" />
Example above is from a project where I only set the minSdkVersion to 23 but nowhere I maintained the targetSdKVersion. So where is that 28 coming from?

**Comment by Equabyte** [comment]
> > So where is that 28 coming from?
> 
> Android `.aar` files can include their own `AndroidManifest.xml` inside. A `.binlog` would help us track this down.
Until updating to 8.0.20, that was being correctly generated as follows:
`<uses-sdk android:minSdkVersion="23" android:targetSdkVersion="34" />`

**Comment by jonathanpeppers** [comment]
@npostma search for `$task ManifestMerger` in the [binlog viewer](https://msbuildlog.com), and see if there is a list of `AndroidManifest.xml` files.

Then I would go inspect them on disk to see which one has `<uses-sdk/>` inside.

Once the rogue `AndroidManifest.xml` is found, we can probably go back to the `.binlog` again to see which `.aar` it came from.

**Comment by npostma** [comment]
![image](https://github.com/dotnet/maui/assets/1578820/23639314-abdd-4bb6-944d-005d1dcbac5d)

A lot of these files are having contents like this (this is the first file)

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="androidx.loader" >

    <uses-sdk
        android:minSdkVersion="14"
        android:targetSdkVersion="28" />

</manifest>
```

**Comment by Equabyte** [comment]
The same issue affects debug builds (same cause: android:targetSdkVersion="28" in manifest file). You can replicate it by uninstalling the app on a Android 14 phone, then deploy again in debug mode. You will see the popup warning "This app was built for an older version of Android".

**Comment by jonathanpeppers** [comment]
If you search for `lp\131` you should be able to find which `.aar` this came from. I don't know off the top of my head what to look for.

If you can share the `.binlog` privately, you can use the email on my GitHub account.

**Comment by npostma** [comment]
Hi Jonathan,

I send you the .binlog file. lp\131 seems to be a/the camera library.

![image](https://github.com/dotnet/maui/assets/1578820/bd5e52f9-91f7-4008-b98f-b69acd0d9b80)

**Comment by jonathanpeppers** [comment]
I'm not able to reproduce the problem with:
* `dotnet new android`
* `dotnet add package Xamarin.AndroidX.Camera.Core --version 1.3.2`
* `dotnet build`
* Open `.apk` in Android Studio

It seems like I should see the issue here:

![image](https://github.com/dotnet/maui/assets/840039/c9fe8942-8fde-4c24-a10e-be38d6d6e9e4)

What do I need to do differently? I'll try a MAUI project and Release mode.

**Comment by jonathanpeppers** [comment]
I tried an _old version_ of the library:

```
dotnet add package Xamarin.AndroidX.Camera.Core --version 1.1.0
```
Also `dotnet new maui` and `Release` mode instead. But it doesn't seem to be happening?

**Comment by Equabyte** [comment]
@jonathanpeppers to replicate the issue with Visual Studio Preview and using 8.0.20 build a MAUI app and deploy it in debug mode on a physical Android device where your app is not installed (uninstall the app prior to deploying in case it was already installed). You should see the popup warning "This app was built for an older version of Android" on your device. Then look at the AndroidManifest.xml file automatically generated and stored inside net8.0-android >: search for the android:targetSdkVersion in that file. What version is it targeting? In my case 28; there seems to be an issue introduced with 8.0.20 due to which the latest SDK version (34) is no longer determined there.

**Comment by jonathanpeppers** [comment]
The MSBuild tasks are implemented in the Android workload in xamarin/xamarin-android, so we should be able to reproduce this without using MAUI at all.

Are you sure it's related to MAUI 8.0.20? I don't even see an update to its AndroidX dependencies in that release:

* https://github.com/dotnet/maui/compare/8.0.14...8.0.20

**Comment by Equabyte** [comment]
> Are you sure it's related to MAUI 8.0.20? I don't even see an update to its AndroidX dependencies in that release:


Would say either there or in the most recent Android SDK34 updates? There was one recently installed in Visual Studio.

**Comment by Equabyte** [comment]
Did a bit of search and in Xamarin there used to be a property AndroidUseLatestPlatformSdk referenced in the build task file Aapt.cs. [See this old entry.](https://github.com/xamarin/xamarin-android/issues/1163) Now the question is, where/how is that property handled nowadays and was there any recent change in that respect?

**Comment by jonathanpeppers** [comment]
No, AndroidUseLatestPlatformSdk has not been used since even early .NET 5 days:

* https://github.com/xamarin/xamarin-android/pull/4453

It was even considered "deprecated" in Xamarin.Android.

**Comment by Equabyte** [comment]
Yep, I realized that and also saw [your proposal for further changes here](https://github.com/dotnet/upgrade-assistant/issues/1508). Anyway: my workaround solution for now is to maintain manually the <uses-sdk/> element in the AndroidManifest file (with minimum and target versions), which works just fine. Somehow VS preview is no longer determining a correct targetSDK automatically matching the compileSDK version in my case. Be it an issue with MAUI 8.0.20 or VS itself, hard for me to tell.

**Comment by npostma** [comment]
@Equabyte and you do not get an error? If i do that get the error that uses-sdk is defined multiple times.

**Comment by Equabyte** [comment]
@npostma no error in my case, and a new version of my app is now in review on Play Store. I edited manually the AndroidManifest.xml file in Visual Studio (edited the file with the xml editor in VS). See element `<uses-sdk` in the app manifest below. 

`<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" android:versionCode="15" android:versionName="2.0.1" package="com.equabyte.DCF_Calc_V1">
	<application android:allowBackup="true" android:supportsRtl="true" android:label="DCF Calc" android:icon="@mipmap/dcfcalcnew1024"></application>
	<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
	<uses-permission android:name="android.permission.INTERNET" />
	<uses-sdk android:minSdkVersion="24" android:targetSdkVersion="34" />
</manifest>`

**Comment by npostma** [comment]
TLDR: Fixed it, i deleted the manifest file and recreated it from scratch. I think something was corrupt. (So i have no uses-sdk in my manifest anymore)

If i changed it or try to add it in the manifest file  i got this error:

```bash
C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.95\tools\Xamarin.Android.Tooling.targets(100,5): error XA1036: AndroidManifest.xml //uses-sdk/@android:minSdkVersion '33' does not match the $(SupportedOSPlatformVersion) value '30.0' in the project file (if there is no $(SupportedOSPlatformVersion) value in the project file, then a default value has been assumed).
1>            C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.95\tools\Xamarin.Android.Tooling.targets(100,5): error XA1036: Either change the value in the AndroidManifest.xml to match the $(SupportedOSPlatformVersion) value, or remove the value in the AndroidManifest.xml (and add a $(SupportedOSPlatformVersion) value to the project file if it doesn't already exist).
1>            C:\src\sc_mc\maui\ActiveID\ActiveID\Platforms\Android\AndroidManifest.xml : warning XA4211: AndroidManifest.xml //uses-sdk/@android:targetSdkVersion '33' is less than $(TargetFrameworkVersion) ''. Using API-34 for ACW compilation.
```

I also tried to install VS2022 preview on an other machine with the android sdk's 33 and 34. To check if its not a bad installation on my main development machine. But a clean install gave me exact the same issues.

So i decided to remove the file and create it from scratch. Now i can make an archive without errors. Also the google console does not gave me errors.

**Comment by jonathanpeppers** [comment]
@npostma can you share your `AndroidManifest.xml` that caused the issue?
